zebedee-reader/src/main/java/com/github/onsdigital/zebedee/search/fastText/FastTextClient.java

```diff
@@ -0,0 +1,186 @@
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+public class FastTextClient implements AutoCloseable {
+
+    private static FastTextClient INSTANCE;
+
+    private final CloseableHttpClient httpClient;
+    private final PoolingHttpClientConnectionManager connectionManager;
+
+    public static FastTextClient getInstance() {
+        if (INSTANCE == null) {
+            synchronized (FastTextClient.class) {
+                if (INSTANCE == null) {
+                    INSTANCE = new FastTextClient(createConfiguration());
+                }
+            }
+        }
+        return INSTANCE;
+    }
+
+    private FastTextClient(ClientConfiguration configuration) {
+        HttpClientBuilder customClientBuilder = HttpClients.custom();
+        this.connectionManager = new PoolingHttpClientConnectionManager();
+        configure(customClientBuilder, configuration);
+        this.httpClient = customClientBuilder.setConnectionManager(this.connectionManager)
+                .build();
+        Runtime.getRuntime().addShutdownHook(new ShutdownHook());
+    }
+
+    private Map<String, Future<BatchSentenceVectorResponse>> submitVectorQueries(Map<String, String> queries,
+                                                                                 FastTextExecutorService executorService) {
+        Map<String, Future<BatchSentenceVectorResponse>> futureMap = new HashMap<>();
+
+        String requestId = UUID.randomUUID().toString();
+        BatchSentenceVectorRequest request = new BatchSentenceVectorRequest(requestId, queries);
+
+        Future<BatchSentenceVectorResponse> future = executorService.submit(request);
+        // Add to futureMap
+        futureMap.put(requestId, future);
+
+        return futureMap;
+    }
+
+    /**
+     * Sets the embedding vector field for a list of pages
+     * @param pages
+     * @param batchThreshold
+     */
+    public List<Page> generateEmbeddingVectors(List<Page> pages, int batchThreshold) throws Exception {
+        // Init a map of uri to page
+        Map<String, Page> pageMap = pages.stream()
+                .collect(Collectors.toMap(x -> x.getUri().toString(), x -> x));
+
+        // Init map to hold batch queries
+        Map<String, String> queries = new HashMap<>();
+
+        // Init map to hold futures
+        Map<String, Future<BatchSentenceVectorResponse>> futureMap = new ConcurrentHashMap<>();
+
+        // Prepare empty embedding vector
+        int dimensions = this.getFastTextInfo().getDimensions();
+        double[] emptyVector = new double[dimensions];
+        String emptyEncodedVector = FastTextHelper.convertArrayToBase64(emptyVector);
+
+        // Build and submit queries
```

Comment:
I feel like this method is quite big - might be worth moving the build & submit part into a separate method.